refactor(mcdu): Flight Phases

MisterChocker opened the pull request:
<!-- Signed-off-by: Leon Beeser <leon.beeser@yahoo.de> -->

<!-- ⚠⚠ Do not delete this pull request template! ⚠⚠ -->
<!-- Pull requests that do not follow this template are likely to be ignored. -->

<!-- Add the issues this PR fixes here. If no issues are related to this PR, then this line can be removed. -->
<!-- Add further issues with a full "Fixes #[issue_no]" line to ensure GitHub closes each one when the PR is merged. -->
<!-- Fixes #[issue_no] -->

## Summary of Changes
<!-- Please provide a summary of changes for this pull request, ensuring all changes are explained. -->

## Screenshots (if necessary)
<!-- If your PR includes visual changes, screenshots from before and after your change should always be included. -->
<!-- Please make best efforts to provide useful before and after screenshots. They should match camera angle, zoom, size, time of day, etc. -->

## References
<!-- You should be making changes based on some kind reference (manuals, videos, IRL photos). P3D/xplane/fsx references will only be accepted if we believe that one cannot reasonably obtain a better source. Please post screenshots of the references you used. Ask around in the discord for how to find references for what you are working on. Exceptions will probably be made for IRL A320 pilots and engineers. -->

<!-- If you are making a pull request related to the MCDU, please make sure you are ONLY referencing the Honeywell Pegasus Step 1A (Rev 0), 2009 edition manual. -->
<!-- If you do not have this manual, please ask on our discord for assistance -->

## Additional context
<!-- Add any other context about the pull request here. -->

<!-- You may optionally provide your discord username, so that we may contact you directly about the issue. -->
Discord username (if different from GitHub):

## Testing instructions
<!-- Detail how this PR should be tested by QA. Try to list important items that need checking, either directly changed by this PR or that could be affected by it -->

<!-- DO NOT DELETE THIS -->
## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created and uploaded.
The build script will have already been run with the latest changes, so no need to rerun it once you download the zip.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the right side, click on the **Artifacts** drop down and click the **A32NX** link

beheh commented on A32NX/html_ui/Pages/A32NX_Core/A32NX_Speeds.js:
```diff
@@ -29,7 +29,7 @@ class A32NX_Speeds {
          */
         setInterval(() => {
-            const fp = Simplane.getCurrentFlightPhase();
+            const fp = SimVar.GetSimVarValue("L:A32NX_FWC_FLIGHT_PHASE", "Enum");
```
This should surely be the FMGC flight phase?

MisterChocker replied:
Yes, it should be, I changed it.

beheh commented on A32NX/html_ui/Pages/A32NX_Core/A32NX_GPWS.js:
```diff
@@ -61,7 +61,7 @@ class A32NX_GPWS {
         const mda = SimVar.GetSimVarValue("L:AIRLINER_MINIMUM_DESCENT_ALTITUDE", "feet");
         const dh = SimVar.GetSimVarValue("L:AIRLINER_DECISION_HEIGHT", "feet");
-        const phase = SimVar.GetSimVarValue("L:AIRLINER_FLIGHT_PHASE", "Enum");
+        const phase = SimVar.GetSimVarValue("L:A32NX_FWC_FLIGHT_PHASE", "Enum");
```
Are you sure this is the FWC flight phase? I'm having a hard time finding a reference for this.

MisterChocker replied:
I've corrected this one as well

beheh commented on A32NX/html_ui/Pages/A32NX_Utils/NXFMGCFlightPhases.js:
```diff
@@ -0,0 +1,10 @@
+const FMGC_FLIGHT_PHASES = {
+    PREFLIGHT: 0,
+    TAKEOFF: 1,
+    CLIMB: 2,
+    CRUISE: 3,
+    DESCENT: 4,
+    APPROACH: 5,
+    GOAROUND : 6,
+    DONE: 7
+};
```
While we do not have real (Typescript) enums here, something irks me about the double uppercase, it looks unusual in the code everywhere.

I think I'd expect either UPPER_CASE.CamelCase, or CamelCase.UPPERCASE. Do we have prior art for other constants like this?

tracernz replied:
`CamelCase.UPPERCASE` looks right to me.

MisterChocker replied:
I'll change that

MisterChocker commented on A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/FMC/A32NX_FMCMainDisplay.js:
```diff
@@ -278,15 +280,17 @@ class FMCMainDisplay extends BaseAirliners {
         if (this.flightPhaseUpdateThrottler.canUpdate(_deltaTime) !== -1) {
-            this.checkUpdateFlightPhase();
+            this.flightPhaseManager.checkFlightPhase();
+            // TODO: remove this.checkUpdateFlightPhase();
+            // this.checkUpdateFlightPhase();
```
Oh yeah, I made this todo to remind me of removing it once I am done, I knew why I made it, will delete.

MisterChocker commented on A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/FMC/A32NX_FMCMainDisplay.js:
```diff
@@ -303,142 +307,61 @@ class FMCMainDisplay extends BaseAirliners {
-                this.currentFlightPhase = FlightPhase.FLIGHT_PHASE_CLIMB;
-                this.climbTransitionGroundAltitude = null;
+        } else if (this.currentFlightPhase === FMGC_FLIGHT_PHASES.CLIMB) {
+            SimVar.SetSimVarValue("L:AIRLINER_V1_SPEED", "Knots", -1);
+            SimVar.SetSimVarValue("L:AIRLINER_VR_SPEED", "Knots", -1);
+            SimVar.SetSimVarValue("L:AIRLINER_V2_SPEED", "Knots", -1);
```
Exactly, when I am not doing this, the PFD Speedband does only display the V2 speed, not selected speed nor managed.

MisterChocker replied:
I'll add a todo to remove that for the new PFD and make a proper cleanup in done phase (as I planned to do (different pr tho)

MisterChocker commented on A32NX/html_ui/Pages/A32NX_Core/FMGC/A32NX_FlightPhaseManager.js:
```diff
@@ -0,0 +1,230 @@
+
+        this.activeFlightPhase.init(_fmc);
+
+        SimVar.SetSimVarValue("L:A32NX_FMGC_FLIGHT_PHASE", "number", FMGC_FLIGHT_PHASES.PREFLIGHT);
```
I'd like to always start in preflight and if needed jump forward.
By triggering the change some stuff gets initialized differently.

beheh replied:
Are you sure about this? Like if you spawn in on an approach, you might already be below the pilot's configured ACC/THR RED altitudes, so how would you ever leave takeoff?

MisterChocker replied:
If you spawn on approach then `this.flightPhaseManager.changeFlightPhase(FmgcFlightPhases.APPROACH)` will be called which initiates the change directly to approach and calls another method which initializes AP changes etc.

tracernz commented on A32NX/html_ui/Pages/A32NX_Core/FMGC/A32NX_FlightPhaseManager.js:
```diff
@@ -0,0 +1,307 @@
+            const planeLla = new LatLongAlt(lat, long);
+            const dist = Avionics.Utils.computeGreatCircleDistance(_fmc.flightPlanManager.decelWaypoint.infos.coordinates, planeLla);
+            if (dist < 3) {
+                _fmc.flightPlanManager._decelReached = true;
```
It's a bit naughty to write another classes members like this.

MisterChocker replied:
yes I don't like it either, eventually this won't be in the flight phase manager, but in the fmgc somewhere itself. possibly getting some inputs from "flight plan manager" or so to call our flight phase change method (which exists for exactly these type of actions)

tracernz replied:
That's cool. Maybe a TODO comment that will highlight in most peoples editors.

MisterChocker replied:
Added todo

beheh commented on A32NX/html_ui/Pages/A32NX_Utils/NXFMGCFlightPhases.js:
```diff
@@ -0,0 +1,10 @@
+const FmgcFlightPhases = {
```
Okay, sorry, to nitpick again on such a minor detail, but:
- I would usually expect an enum-ish object to have the singular name (`FlightPhase`)
- Fmgc is fine... however I'm personally not that strict with CamelCase so I'm personally also okay with FMGCFlightPhase, but maybe more people should chime in or voice their preference here.

MisterChocker replied:
Yeah I personally went with camel case, but since FMGC is a name itself, we might want to stick with that.
I'd like to get peoples thoughts on this one.

MisterChocker replied:
@tracernz thoughts?

tracernz replied:
There are good arguments both ways re. FMGC vs. Fmgc and I don't really mind either. I think having FMGC in the name is a good move to avoid confusion with FWC or any other systems. It's easier to spot issue in future code review, e.g. if you see FmgcFlightPhase in some ECAM code you'd immediately be suspicious, whereas FlightPhase would seem normal on first inspection.

beheh commented on A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_PerformancePage.js:
```diff
@@ -994,11 +994,11 @@ class CDUPerformancePage {
         ]);
     }
-    static UpdateThrRedAccFromOrigin(mcdu, updateThrRedAlt = true, updateAccAlt = true) {
+    static UpdateThrRedAccFromOrigin(mcdu, updateThrRedAlt = true, updateAccAlt = true, autoRefresh = false) {
```
Hm, this is a bit different than it currently works. Right the it's the responsibility of the thing calling `UpdateThrRedAccFromOrigin` to ensure it doesn't overwrite pilot input, hence `updateThrRedAlt` and `updateAccAlt`. If you move this check inside the component you don't really need to specify what you're updating, you can just call this whenever.

MisterChocker replied:
I don't really understand what you are trying to say with ` you don't really need to specify what you're updating, you can just call this whenever.`

MisterChocker commented on A32NX/html_ui/Pages/A32NX_Core/FMGC/A32NX_FlightPhaseManager.js:
```diff
@@ -0,0 +1,307 @@
+        }
+    }
+
+    handleFcuInput() {
```
Flight Phases only react to alt knob push/pull, so I find that naming perfectly fine.

tracernz replied:
I'm fine with either way so leave as is and mark this resolved if you like. I try to look at APIs on their own merit rather than from the lens of a specific consumer, even if it seems obvious now that there wouldn't be any new consumers later.

MisterChocker commented on A32NX/html_ui/Pages/A32NX_Core/FMGC/A32NX_FlightPhaseManager.js:
```diff
@@ -0,0 +1,307 @@
+        // Try initiate des
+        if (
+            this.fmc.currentFlightPhase !== FmgcFlightPhases.DESCENT &&
+            (this.fmc.flightPlanManager.getDestination()).liveDistanceTo < 200 &&
```
Fixed

MisterChocker commented on A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_MainDisplay.js:
```diff
@@ -233,8 +233,7 @@ class A320_Neo_CDU_MainDisplay extends FMCMainDisplay {
         if (initialFlightPhase) {
-            this.currentFlightPhase = initialFlightPhase;
-            this.onFlightPhaseChanged();
+            this.flightPhaseManager.changeFlightPhase(initialFlightPhase === FlightPhase.FLIGHT_PHASE_TAXI ? FMGC_FLIGHT_PHASES.PREFLIGHT : initialFlightPhase);
```
Done

beheh commented on A32NX/html_ui/Pages/A32NX_Core/FMGC/A32NX_FlightPhaseManager.js:
```diff
@@ -131,15 +131,26 @@ class A32NX_FlightPhaseManager {
+    check(_deltaTime, _fmc) {
         // Simplane.getAltitudeAboveGround() > 1500; is used to skip flight phase in case ac reloaded midair
-        return Simplane.getEngineThrottleMode(0) >= ThrottleMode.FLEX_MCT || Simplane.getEngineThrottleMode(1) >= ThrottleMode.FLEX_MCT || Simplane.getAltitudeAboveGround() > 1500;
+        return this.takeoffConfirmation.write((
+            (
+                Simplane.getEngineThrottleMode(0) >= ThrottleMode.FLEX_MCT ||
+                Simplane.getEngineThrottleMode(1) >= ThrottleMode.FLEX_MCT
+            ) && (
+                SimVar.GetSimVarValue("ENG N1 RPM:1", "percent") > .75 ||
+                SimVar.GetSimVarValue("ENG N1 RPM:2", "percent") > .75
+            ))
+        || Simplane.getAltitudeAboveGround() > 1500,
+        _deltaTime
+        );
```
Personally I find it more readable like this (prettier-style):

```suggestion
        return this.takeoffConfirmation.write(
            (
                (
                    Simplane.getEngineThrottleMode(0) >= ThrottleMode.FLEX_MCT ||
                    Simplane.getEngineThrottleMode(1) >= ThrottleMode.FLEX_MCT
                ) && (
                    SimVar.GetSimVarValue("ENG N1 RPM:1", "percent") > .75 ||
                    SimVar.GetSimVarValue("ENG N1 RPM:2", "percent") > .75
                )
            ) || Simplane.getAltitudeAboveGround() > 1500,
            _deltaTime
        );
```

St54Kevin commented:
Category: QA Tester
Name: St54Kevin#3583
Date of testing: 01/03/2021
Version of the sim: 1.13.16.0
PR Tested: #3658 
Tier of testing: 2
Changes to observe: Flight Phases
Testing technique: 
Steps you've made to test the Pull Request:

1. Made a short VATSIM flight LOWW-EDDP
2. Levelled off 200ft below CRZ ALT, observed FMA ALT, changed CRZ ALT on PROG page, observed FMA ALT CRZ
3. Step climb, observed stay in CRZ phase
4. Normal descent
5. Had to activate appr phase manually as the plane didnt flew the STAR
6. Go around and second approach (did not work, see issues)
7. Tried a direct turnaround
8. Waited 15min but the MCDU did not reset
9. Tried just loading a new simbrief flightplan, but did also not work
![image](https://user-images.githubusercontent.com/448627/109482384-38106a80-7a7e-11eb-99dd-384f2e053909.png)
![image](https://user-images.githubusercontent.com/448627/109482407-3fd00f00-7a7e-11eb-95ca-bb8201203251.png)


Media:
On descent the ALTs on Go AROUND Page were empty, they shouldnt. Is that related?
![image](https://user-images.githubusercontent.com/448627/109476236-f203d880-7a76-11eb-8f87-8421f165d43e.png)


Issues (If Any): When levelling off at CRZ ALT, the CRZ field on the PROG page will turn to white dashes, I think it should keep the blue input CRZ level. I then could not put in a new FL, as it says "ENTRY OUT OF RANGE". Is that related to this PR? 
Go around phase did not activate at TOGA. Nothing changed, plane stayed in GS/LOC, no SRS commands nothing. Had to do it all manually. it went to DES again before second approach, but then in approach phase managed wasnt working, it commanded 250kts.

Overall Rating: LGTM

Conclusions of the testing:

MisterChocker commented:
@St54Kevin thanks for noticing, somehow that slipped through.
I've found the troublemakers but still need to do some tests myself.

MisterChocker commented:
@St54Kevin Looks like I now have managed to fix all things you have noted.

I would appreciate a retest.

FYI As of now you are not able to exit the go around phase except via entering the approach phase.

FlightSimmer7700 commented:
Category: QA Tester
Name: FlightSimmer7700#3468
Date of testing: 2/3/21
Version of the sim: 1.13.17.0
PR Tested: #3658 
Tier of testing: 2
Changes to observe:
Flight phases

Testing technique: Did a full flight
Steps you've made to test the Pull Request:
1. Loaded in 
2. Did a step climb
3. descended during cruise (360) to 280 and back up
4. Flew to destination
5. Did a go-around
6. landed
7. did a turnaround up to 15000ft

Media: Insert screenshots, recordings, or anything relevant to the review here.
Too large for git: [](https://discord.com/channels/738864299392630914/754130199804772372/816453178714882089)

![](https://cdn.discordapp.com/attachments/732004815453945916/816436955020787753/unknown.png)

![](https://cdn.discordapp.com/attachments/732004815453945916/816445559765336084/unknown.png)

![](https://cdn.discordapp.com/attachments/732004815453945916/816448638443061288/unknown.png)


Issues (If Any): 

- After descending more than 200nm away from the airport, when climbing back to your cruise the aircraft goes into climb phase but when coming down it stayed in cruise (I guess that is how it is supposed to work)(when back to cruise alt it goes back into the cruise flight phase as expected)
- When descending after doing the above (not sure if it happens without) it stays in the cruise flight phase until you manually set approach
- When doing a turnaround the TO and Approach info from the previous flight stay in (not sure but I would think it would get removed automatically once the route is changed)

Steps to achieve the issue: do a flight with the steps included above

Overall Rating: To be improved
Conclusions of the testing: some improvements to be made but good work

MisterChocker commented:
@FlightSimmer7700
- when you start a descent more then 200 nm away from your destination, you will maintain in cruise phase.
(see instructions)
- this pr does not add/do a full reset of the fmgc, it just introduces proper flight phase management.

FlightSimmer7700 commented:
> @FlightSimmer7700
> 
> * when you start a descent more then 200 nm away from your destination, you will maintain in cruise phase.
>   (see instructions)
> * this pr does not add/do a full reset of the fmgc, it just introduces proper flight phase management.

Ok so it should have not gone into the climb phase when I climbed back up?

beheh commented:
> Perform an early descent (more than 200nm before destination, AC should not go into descent phase)

I'm not convinced about that:
![image](https://user-images.githubusercontent.com/929769/109792443-d08c2380-7c13-11eb-91ec-f0079f3f8761.png)

This is an OR. Isn't the first condition met?

In fact, further on it's explicitly called out how to enter DESCENT from more than 200 NM out:

![image](https://user-images.githubusercontent.com/929769/109792560-eac60180-7c13-11eb-8419-2bfddd6ff743.png)

We may need more sources on this?

MisterChocker commented:
Oh now I see. I'll get to it.

MisterChocker commented:
I've fixed the descent initiation conditions:
These are only triggered on fcu alt knob pull or push.
Now are:
- descent:
   - not in descent flight phase
   - AND distance is less than 200nm to destination OR destination is not set
   - AND FCU selected altitude is below cruise altitude
- early descent:
   - not in descent flight phase
   - AND distance is less than 200nm to destination
   - AND AC is above FL200 AND FCU target altitude is below FL200

St54Kevin commented:
> I've fixed the descent initiation conditions:
> (These are only triggered on fcu alt knob pull or push.)
> Now are:
> - descent:
>    - not in descent flight phase
>    - AND distance is less than 200nm to destination OR destination is not set
>    - AND FCU selected altitude is below cruise altitude
> - early descent:
>    - not in descent flight phase
>    - AND distance is less than 200nm to destination
>    - AND AC is above FL200 AND FCU target altitude is below FL200

Isn't that wrong? There are no ANDs in Behehs reference above I think. It goes into descent anytime you will initiate it to go below cruise ALT. Of course you can reenter cruise phase by lowering the PROG page CRZ alt

MisterChocker commented:
@St54Kevin how would that work when these aren't ands?
Just being in range of 200nm or further aways does not initiate a transition into descent.
And the fcu selected altitude has to be lower than the cruise altitude, otherwise a step climb will be initiated or nothing happens at all.

beheh commented:
I think @MisterChocker's suggested version looks fine, the OR is on the top level, the ANDs are only between things that do belong together.

marcman86 commented:
Category: QA Tester
Name: marcman86#4907
Date of testing: 5.3.2021
Version of the sim: 1.13.17.0
PR Tested: #3658
Tier of testing: 2
Changes to observe: Flight phases
Testing technique: Tryout

Steps made to test the Pull Request:

CRZ level is FL200

1.) Spawned C&D at EDDP (Take off white)
2.) Took off to FL200 (Take off green, CLB, CRZ)

Now performed >200NM to destination:

3.) Open descent to FL180 (CRZ)
4.) Open climb to FL200 (CLB, CRZ)
5.) Open climb to FL220 (CRZ, ALT CRZ)
6.) Open descent to FL180 (DES, ALT)
7.) Open climb to FL200 (CLB, ALT)

Here it stays in CLB phase.

7.) Open climb to FL210 (CLB, ALT)
8.) Open descent to FL200 (CLB, ALT)

Still in CLB phase.

Issues: Don't go back to CRZ after early descent and CLB again.

Overall Rating: Needs to be improved

Conclusions of the testing: Think i found something.

Hope this helps :-)

MisterChocker commented:
Yes, needed logic has not been implemented yet (I'll push it shorty)

During Climb Phase:
When no cruise alt is defined (e.g. after descent) the cruise flight level default's to the fcu selected alt

marcman86 commented:
Category: QA Tester
Name: marcman86#4907
Date of testing: 6.3.2021
Version of the sim: 1.13.17.0
PR Tested: #3658
Tier of testing: 2
Changes to observe: Flight phases
Testing technique: Tryout

Steps made to test the Pull Request:

CRZ level is FL190

1.) Spawned C&D at EDDB (Take off white)
2.) Took off to FL190 (Take off green, CLB, CRZ)

Now performed >200NM to destination:

3.) Open descent to FL180 (CRZ), PFD shows ALT CRZ (?)
4.) Open climb to FL190 (CLB, CRZ), PFD shows ALT CRZ (!)
5.) Open climb to FL210 (CRZ), PFD shows ALT CRZ
6.) Open descent to FL180 (DES), PFD shows ALT
7.) Open climb to FL190 (CLB, CRZ), PFD shows ALT CRZ (!)
8.) Open descent to FL180 (CRZ), PFD shows ALT CRZ (?)
9.) Open climb to FL190 (CLB), on 18500 descent to FL180 (CLB), PFD shows ALT
10.) Open climb to FL190 (CLB, CRZ), PFD shows ALT CRZ (!)
11.) Open climb to FL200 (CLB, CRZ), PFD shows ALT CRZ

Stuck in CLB seems to be fixed.

To simulate a Go Around:
12.) Open descent to FL100 (CRZ), PFD shows ALT CRZ
13.) Manual activate APP phase (APPR)
14.) Set throttle lever to TOGA (GO AROUND)
15.) Open climb to FL190 (Stuck in GO AROUND, as stated in the description)

Issues (don't know if part of this PR, also please check if this is true):

- ALT CRZ should only show up > CRZ lvl
- ALT should show up < CRZ lvl
- CRZ should show up on CRZ lvl

Overall Rating: Needs to be improved

Conclusions of the testing: Will do a second test (if i find the time) to do a normal flight with a turn around.

Hope this helps :-)

MisterChocker commented:
@marcman86 The FMA seems to have it's own issues (unrelated to this pr).

When initiating a descent the crz fl field on the prog page displays dashes, I assumed that this means there is no longer a crz fl.
When in climb and no crz fl is entered, the crz fl field defaults to the fcu selected altitude.

This is from the honeywell manual.

MisterChocker commented:
Alright, I've fixed a couple of things:
- AC stays in cruise phase when doing step climbs,
- AC does not change flight phases when pushing/pulling alt know except for DES initiation,

and these are new additions:
- On TakeOff (alt knob push/pull initiates CLB phase),
- On Approach: insert valid a cruise alt in 1L on prog page to re-establish climb phase,
- On GoAround: insert valid a cruise alt in 1L on prog page to re-establish climb phase.

Added two more conditions for approach activation:
Now: Decel Point passed AND AGL < 9500ft AND (in NAV MODE OR LOC* or LOC engaged)

marcman86 commented:
Category: QA Tester
Name: marcman86#4907
Date of testing: 6.3.2021
Version of the sim: 1.13.17.0
PR Tested: #3658
Tier of testing: 2
Changes to observe: Flight phases
Testing technique: Tryout

Steps made to test the Pull Request:

CRZ level is FL210

1.) Spawned on runway at EDDP (TAKE OFF RWY white)
2.) Took off to FL210 (TAKE OFF RWY green, CLB, CRZ), PFD shows CRZ
Now performed >200NM to destination:
3.) Open climb to FL220 (CRZ), PFD shows ALT CRZ
4.) Open descent to FL210 (CRZ), PFD shows ALT CRZ (!) (CRZ lvl is still set in Init and Prog)
5.) Open descent to FL205 (CRZ), PFD shows ALT CRZ (?) 
6.) Open climb to FL210 (CRZ), PFD shows ALT CRZ (!) (CRZ lvl is still set in Init and Prog)
7.) Open descent to FL190 (DES), PFD shows ALT
8.) Open climb to FL210 (DES), PFD shows ALT (CRZ lvl is still set in Init, but not Prog. Not able to reenter it)
9.) Open descent to 5000ft (DES), PFD shows ALT
10.) Activate APPR phase, set CRZ lvl on Prog to 6000ft
11.) Open climb to 6000ft (CLB,CRZ), PFD shows ALT CRZ (!) (CRZ lvl 6000ft shown in Init and Prog)
12.) Open descent to 3000ft (CRZ), PFD shows ALT CRZ (!) (CRZ lvl 6000ft shown in Init and Prog CRZ lvl changes with set FCU level) (!)
13.) Activate APPR phase, auto reset to CRZ cause issue of 12. Need to set Prog CRZ lvl to FL200 to force the AC to stay in APPR phase
14.) Get the AC 1000ft AGL and set APPR phase
15.) Set lever to TOGA to get into Go Around phase
16.) Set FL210 in Prog and open climb to FL210 (CLB, CRZ)
17.) Open descent FL205 (CRZ), PFD shows ALT CRZ (!) (CRZ lvl 6000ft shown in Init and Prog CRZ lvl changes with set FCU level) (!)
18.) Open climb to FL210 (CRZ), PFD shows ALT CRZ (!)
Now performed <200NM to destination:
19.) Open descent FL205 (DES), PFD shows ALT
20.) Open descent FL110 (DES), PFD shows ALT
21.) On DECEL AC goes in APPR phase
22.) Land EGLL
23.) MCDU resets on stop at runway
24.) Set EGLL/EDDP goes to TAKE OFF RWY white

Issues: See 12. and 17.

Overall Rating: Needs to be improved

Conclusions of the testing: Hard to test

Hope this helps :-)

marcman86 commented:
Category: QA Tester
Name: marcman86#4907
Date of testing: 6.3.2021
Version of the sim: 1.13.17.0
PR Tested: #3658
Tier of testing: 1
Changes to observe: Flight phases
Testing technique: Tryout

Steps made to test the Pull Request:

Did the flight like before (without landing).

CRZ lvl on Prog changes due to step climb and step descent. This was intentional.

Rechecked the CRZ > APPR behaviour. Now the AC stays in APPR phase.

PFD Alt displays is not part of this PR.

Issues: -

Overall Rating: Good

Conclusions of the testing: Seems to work.

Hope this helps :-)

St54Kevin commented:
Category: QA Tester
Name: St54Kevin#3583
Date of testing: 08/03/2021
Version of the sim: 1.13.17.0
PR Tested: #3658 
Tier of testing: 2
Changes to observe: Flight phases

Steps you've made to test the Pull Request:

**Full flight LOWW-EDDH and back**
1. Climb normal - TO phase at FLEX, CLB phase at ACC ALT
2. Level off at FL360 (PROG CRZ set to FL380) - CLB phase
3. changed PROG CRZ to FL360 - CRZ phase
4. step climb to FL380 - "NEW CRZ ALT" - CRZ phase
5. Started descent 170nm from dest (FPLN distance) - DES phase
6. All NAV and managed, over DECEL point - auto APPR phase
7. Had to change RWY due to ATC, vectors to new approach
8. Normal landing - After vacating TO Page shown again **<- is this right?**
9. Turnaround, ACARS loading worked this time

10. Second flight back to LOWW exactly like the first one, no issues
11. Go around in VIE - @TOGA GoAround phase
13. Approach page was filled out with data from previous flight
14. manual APPR activation - APPR phase
15. normal landing

Media: My reference:
![rsz_image0](https://user-images.githubusercontent.com/448627/110312634-a10b5b80-8005-11eb-84e5-9ae5b4711503.jpg)

Issues (If Any): Points 8. and 13. Think this will get fixed after this PR

Conclusions of the testing: Very well done. All the problems are solved. I could not find any hickups

MisterChocker commented:
Before merging I do like to look into the issue after landing where performance page was shown again
